include/mata/sparse-set.hh

```diff
@@ -0,0 +1,355 @@
+        void erase(InputIterator first, InputIterator last) {
+            while (first != last) {
+                erase(*first);
+                first++;
+            }
+            assert(consistent());
+        }
+
+        template<class T>
+        void erase_all(const T & set) {
+            erase(set.begin(),set.end());
+        }
+
+        void erase_all(const std::initializer_list<Number> & list) {
+            erase(list.begin(),list.end());
+        }
+
+        //call this (instead of the fried are_disjoint) if you want the other container to be iterated (e.g. if it does not have constant membership)
+        template<class T>
+        bool intersects_with(const T & set) const {
+            for (auto i=set.begin();i<set.end();i++) {
+                if (has(*i))
+                    return true;
+            }
+            return false;
+        }
+
+        /*
+         * Complements the set with respect to a given number of elements = the maximum number + 1.
+         */
+        //probably can be done little faster
+        void complement(Number domain_size) {
+            Number orig_capacity = capacity_;
+            for (Number i = 0; i<domain_size; ++i) {
+                if (has(i))
+                    erase(i);
+                else
+                    insert(i);
+            }
+            for (Number i = domain_size; i<orig_capacity; ++i) {
+                erase(i);
+            }
+
+            assert(consistent());
+        }
+
+        template<typename F>
+        void filter(F && is_staying) {
+            for (Number i = 0; i < size_; i++) {
+                while (i<size_ && !is_staying(dense[i])) {
+                    erase(dense[i]);
+                }
+            }
+
+            assert(consistent());
+        }
+
+        void sort() {
+            std::sort(dense.begin(),dense.begin()+size_);
+            for (Number i = 0; i<size_; i++) {
+                sparse[dense[i]] = i;
+            }
+
+            assert(consistent());
+        }
+
+        template<typename F>
+        void rename(F && renaming) {
+            for (Number i = 0; i < size_; i++) {
+                if (dense[i] != renaming(dense[i])) {
+                    dense[i] = renaming(dense[i]);
+                    if (dense[i] >= capacity_)
+                        reserve(dense[i]+1);
+                    sparse[dense[i]] = i;
+                }
+            }
+
+            assert(consistent());
+        }
+
+        Number max() {
+            Number max = 0;
+            for (Number i = 0;i<size_;i++) {
+                if (max < dense[i])
+                    max = dense[i];
+            }
+            return max;
+        }
+
+        void truncate() {
+            if (size_ == 0)
+                capacity_ = 0;
+            else
+                capacity_ = max()+1;
+
+            assert(consistent());
+        }
+
+        // This could be a template with types of A and B both as parameters, in Util?
```

Comment:
We would have to unify the membership check function names. `has()` does not work for anything other than sparse set. I would personally refrain from moving it to `Nfa::Util`.

Reply:
with contains? Or we could use [].

Reply:
You could have more success with `operator[]`. I still believe this would not improve anything for us, though.

What do others think?